### PR TITLE
Create ansible role Certbot

### PR DESCRIPTION
**Pull request by lsaintier**
Installs certbot, and configures a cronjob for certificate renew.

Compatible with http-01 and dns-01 challenge types.

When using http-01 challenge, this role is only compatible with webroot plugin. Using other plugins would break idempotency with other middlewares configured with ansible.

### CODE REVIEW

**Comment by geetoodeetoo** [comment]
@ElieDeloumeau  we need a review ;-)

File: vars/main.yml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+# vars file for role
+ansible_distro: "{{ ansible_distribution|lower|replace('\"', '') }}"
+ansible_distro_version: "{{ ansible_distribution_version|lower|replace('/', '_') }}"
+ansible_distro_release: "{{ ansible_distribution_release|lower }}"
+ansible_distro_major_version: "{{ ansible_distribution_major_version|lower|replace('/', '_') }}"
```
Comment (ElieDeloumeau) [inline vars/main.yml:6]: These variables are not used, you should remove this file

File: tasks/create-cert-webroot.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+---
+- name: Check if certificate already exists.
+  stat:
+    path: /etc/letsencrypt/live/{{ cert_item.domains | first }}/cert.pem
+  register: letsencrypt_cert
```
Comment (ElieDeloumeau) [inline tasks/create-cert-webroot.yml:5]:
```suggestion
  register: _certbot_letsencrypt_cert
```

Private variables should be prefixed with `_certbot_`

File: tasks/create-cert-webroot.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+- name: Create webroot directory if it doesn't exist yet
+  file:
+    path: "{{ cert_item.webroot | default(certbot_webroot) }}"
+    state: directory
+    group: www-data
+    mode: 0750
+
+- name: Generate new certificate if one doesn't exist, one host at a time
+  command: "{{ certbot_create_command }}"
+  register: generate_certificate
```
Comment (ElieDeloumeau) [inline tasks/create-cert-webroot.yml:16]:
```suggestion
  register: _certbot_generate_certificate
```

Private variables should be prefixed with `_certbot_`

File: tasks/create-cert-webroot.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    msg:
+      - "****"
+      - "**** !!! MANUAL RELOAD NEEDED !!!"
+      - "**** Configuration files changed, you have to..."
+      - "**** 1. update your webserver configuration to use:"
+      - "**** 1.1. certificate in /etc/letsencrypt/live/{{ cert_item.domains | first }}/cert.pem"
+      - "**** 1.2. private key in /etc/letsencrypt/live/{{ cert_item.domains | first }}/privkey.pem"
+      - "**** 2. and reload it"
+      - "****"
+  when: generate_certificate.changed
```
Comment (ElieDeloumeau) [inline tasks/create-cert-webroot.yml:30]:
```suggestion
  when: _certbot_generate_certificate.changed
```

Private variables should be prefixed with `_certbot_`

File: tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        - "{{ ansible_distro }}-{{ ansible_distro_major_version }}.yml"
+        - "{{ ansible_distro }}.yml"
+        - "{{ ansible_os_family|lower }}-family.yml"
+        - "{{ ansible_system | lower }}.yml"
+      paths:
+        - "vars"
+      skip: true
+  loop_control:
+    loop_var: loop_vars
+
```
Comment (ElieDeloumeau) [inline tasks/main.yml:19]:
```suggestion
```

This code is not used, so you can should delete it

File: tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    - certbot_reload_services | length > 0
+    - certbot_reload_services_before_enabled
+
+- include_tasks: install-certbot.yml
+  tags: certbot_install
+
+- include_tasks: create-cert-webroot.yml
+  with_items: "{{ certbot_certs }}"
+  loop_control:
+    loop_var: cert_item
```
Comment (ElieDeloumeau) [inline tasks/main.yml:32]:
```suggestion
    loop_var: _certbot_cert_item
```

Private variables should be prefixed with `_certbot_`

File: tasks/create-cert-webroot.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
+---
+- name: Check if certificate already exists.
+  stat:
+    path: /etc/letsencrypt/live/{{ cert_item.domains | first }}/cert.pem
```
Comment (ElieDeloumeau) [inline tasks/create-cert-webroot.yml:4]:
```suggestion
    path: /etc/letsencrypt/live/{{ _certbot_cert_item.domains | first }}/cert.pem
```

Private variables should be prefixed with `_certbot_`

File: tasks/create-cert-webroot.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+  register: letsencrypt_cert
+
+- name: Create webroot directory if it doesn't exist yet
+  file:
+    path: "{{ cert_item.webroot | default(certbot_webroot) }}"
```
Comment (ElieDeloumeau) [inline tasks/create-cert-webroot.yml:9]:
```suggestion
    path: "{{ _certbot_cert_item.webroot | default(certbot_webroot) }}"
```

Private variables should be prefixed with `_certbot_`

File: tasks/install-certbot.yml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+---
+- name: Install Certbot from repository and related packages
+  package:
+    name: "{{ item }}"
+    state: "present"
+  loop: "{{ certbot_packages }}"
```
Comment (ElieDeloumeau) [inline tasks/install-certbot.yml:6]:
```suggestion
- name: Install Certbot from repository and related packages
  package:
    name: "{{ certbot_packages }}"
    state: "present"
```

File: tasks/install-certbot.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+
+- name: Install Certbot plugin
+  pip:
+    name: "{{ item }}"
+  loop: "{{ certbot_plugins }}"
```
Comment (ElieDeloumeau) [inline tasks/install-certbot.yml:11]:
```suggestion
- name: Install Certbot plugins
  pip:
    name: "{{ certbot_plugins }}"
```

File: tasks/install-certbot.yml [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+- name: Install cron
+  apt: name=cron state=present
```
Comment (ElieDeloumeau) [inline tasks/install-certbot.yml:14]:
```suggestion
- name: Install cron
  package:
    name: cron
    state: present
```

Non OS specific task

**Review by ElieDeloumeau — CHANGES_REQUESTED** [review]
Could you please test it on Ubuntu, CentOS and AmazonLinux? I think it can work as is

**Comment by rsicart** [comment]
> Could you please test it on Ubuntu, CentOS and AmazonLinux? I think it can work as is

Ubuntu is OK, but for CentOS and AmazonLinux there's more work to do.

Perhaps we can remove CentOS 8 from tests https://github.com/ansible-collections/news-for-maintainers/discussions/4 ?

**Comment by ElieDeloumeau** [comment]
> Perhaps we can remove CentOS 8 from tests [ansible-collections/news-for-maintainers#4](https://github.com/ansible-collections/news-for-maintainers/discussions/4) ?

CentOS was to validate the RHEL compatibility but indeed we need to replace this test with [UBI](https://www.redhat.com/en/blog/introducing-red-hat-universal-base-image) later.

We can skip the compatibility with RHEL and AmazonLinux for now, I'll create a new PR later.

**Review by ElieDeloumeau — APPROVED** [review]

**Pull request merged**